internal/cloud/realm/client.go

```diff
@@ -91,5 +93,52 @@ func (c *client) do(method, path string, options api.RequestOptions) (*http.Resp
 		req.Header.Set(api.HeaderAuthorization, "Bearer "+auth)
 	}
 	client := &http.Client{}
-	return client.Do(req)
+
+	res, err := client.Do(req)
+	if res.StatusCode != http.StatusUnauthorized || !options.UseAuth {
+		return res, err
+	}
+
+	authToken, refreshErr := c.refreshAuth(client)
+	if refreshErr != nil {
+		return nil, refreshErr
+	}
+	req.Header.Set(api.HeaderAuthorization, "Bearer "+authToken)
+	// TODO REALMC-7580 change these lines to just return client.Do(req) as session will be cleared downstream
+	res, err = client.Do(req)
+	if res.StatusCode == http.StatusUnauthorized {
+		c.profile.ClearSession()
+		return nil, err
+	}
+	return res, err
+}
```

Comment:
I had the code this way, instead of another method on client because it felt like the simplest approach for avoiding an infinite loop of retries without having to pass in some information to know to stop retrying if the retry call keeps giving 401s